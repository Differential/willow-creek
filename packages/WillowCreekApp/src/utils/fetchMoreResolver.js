import get from 'lodash/get';
import set from 'lodash/fp/set';

export default function fetchMoreResolver({
  collectionName,
  variables,
  fetchMore,
  data,
} = {}) {
  return () => {
    const pageInfoPath = `${collectionName}.pageInfo`;
    const edgePath = `${collectionName}.edges`;

    const after = get(data, `${pageInfoPath}.endCursor`);
    if (!after) return;

    fetchMore({
      variables: { ...variables, after },
      updateQuery: (previousResult, { fetchMoreResult }) => {
        // combine previous data and fetchMore data
        const newDataWithPageInfo = set(
          pageInfoPath,
          {
            ...get(previousResult, pageInfoPath, {}),
            ...get(fetchMoreResult, pageInfoPath, {}),
          },
          previousResult
        );
        const newDataWithAdditionalItems = set(
          edgePath,
          [
            ...get(previousResult, edgePath, []),
            ...get(fetchMoreResult, edgePath, []),
          ],
          newDataWithPageInfo
        );

        return newDataWithAdditionalItems;
      },
    });
  };
}
