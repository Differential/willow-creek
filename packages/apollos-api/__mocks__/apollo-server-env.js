import fetch from 'jest-fetch-mock';

import * as apolloDatasourceMocks from './apollo-datasource-mocks';
import * as rockMocks from './rock-api-mocks';

const apolloServerEnv = require.requireActual('apollo-server-env');

// eslint-disable-next-line
const Response = apolloServerEnv.Response;

const resolveWith = (data, url) =>
  Promise.resolve(
    new Response(JSON.stringify(data), {
      url,
      status: 200,
      statusText: 'OK',
      headers: new apolloServerEnv.Headers({
        'Content-Type': 'application/json',
      }),
    })
  );

fetch.mockLiveDataSourceApis = () => {
  fetch.mockImplementation((request) => {
    if (request.url.match('/api/v1/events/current')) {
      return resolveWith(apolloDatasourceMocks.liveStreamLive(), request.url);
    }
    if (request.url.match('api.scripture.api.bible')) {
      return resolveWith(apolloDatasourceMocks.Scripture(), request.url);
    }
    return Promise.reject();
  });
};

fetch.mockRockDataSourceAPI = () => {
  fetch.mockImplementation((request) => {
    let { url } = request;
    url = decodeURI(url);
    if (!url.match('https://apollosrock.newspring.cc/api')) {
      if (request.url.match('/api.scripture.api.bible/v1')) {
        return resolveWith(apolloDatasourceMocks.Scripture());
      }
      return Promise.reject();
    }

    if (url.match('api/EntityTypes')) {
      return resolveWith(
        [
          {
            Id: 201,
          },
        ],
        url
      );
    }

    if (url.match('api/InteractionChannels')) {
      return resolveWith(
        [
          {
            Id: 10,
          },
        ],
        url
      );
    }

    if (url.match('api/InteractionComponents')) {
      return resolveWith(
        [
          {
            Id: 101,
          },
        ],
        url
      );
    }

    if (url.match('api/ContentChannels/\\d')) {
      return resolveWith(rockMocks.contentChannel());
    }

    if (url.match('api/Groups/GetFamilies/\\d')) {
      return resolveWith(rockMocks.families());
    }

    if (url.match('api/ContentChannels')) {
      return resolveWith([rockMocks.contentChannel()]);
    }

    if (url.match('api/ContentChannelItems/test-case-no-attributes')) {
      const mock = rockMocks.contentItem();
      mock.AttributeValues = {};
      mock.Attributes = {};
      return resolveWith(mock);
    }

    if (url.match('api/ContentChannelItems/123')) {
      const contentItem = rockMocks.contentItem();
      contentItem.AttributeValues.Scriptures = { Value: 'John 3:16' };
      return resolveWith(contentItem);
    }

    if (url.match('api/ContentChannelItems/\\d')) {
      return resolveWith(rockMocks.contentItem());
    }

    if (url.match('api/ContentChannelItems')) {
      return resolveWith([rockMocks.contentItem()]);
    }

    if (url.match('api/ContentChannelItemAssociations')) {
      return resolveWith([rockMocks.contentChannelItemAssociation()]);
    }

    if (url.match('api/Auth/Login')) {
      const body = JSON.parse(request.body);
      const response = new Response('');
      if (body.Password === 'good') {
        response.headers.set('set-cookie', 'some cookie');
        return Promise.resolve(response);
      }
      return Promise.reject({ ...response, status: 401 });
    }

    if (url.match('api/People/GetCurrentPerson')) {
      return resolveWith(rockMocks.people());
    }

    if (url.match('api/People/\\d')) {
      return resolveWith(rockMocks.people());
    }

    if (url.match('api/People')) {
      if (request.method === 'POST') {
        const { Email } = JSON.parse(request.body);
        if (!Email) {
          const response = new Response('');
          response.status = 400;
          return Promise.reject(response);
        }
        return resolveWith('35');
      }

      return resolveWith([rockMocks.people()]);
    }

    if (url.match('api/Interactions/\\d')) {
      return resolveWith(rockMocks.interaction());
    }

    if (url.match('api/Interactions')) {
      if (request.method === 'GET') {
        return resolveWith([rockMocks.interaction()]);
      }
      if (request.method === 'POST') {
        const response = new Response('456', {
          status: 200,
          statusText: 'OK',
          headers: new apolloServerEnv.Headers({
            'Content-Type': 'application/json',
          }),
        });
        return Promise.resolve(response);
      }
    }

    if (url.match('api/InteractionSessions')) {
      if (request.method === 'GET') {
        return resolveWith(rockMocks.session());
      }
      if (request.method === 'POST') {
        const response = new Response('123', {
          status: 200,
          statusText: 'OK',
          headers: new apolloServerEnv.Headers({
            'Content-Type': 'application/json',
          }),
        });
        return Promise.resolve(response);
      }
    }

    if (url.match('api/UserLogins')) {
      if (request.method === 'POST') {
        const { UserName } = JSON.parse(request.body);
        if (!UserName) {
          const response = new Response('');
          response.status = 400;
          return Promise.reject(response);
        }
        return resolveWith({ id: 21 });
      }
      const identity = url // identity = UserName
        .split('eq')
        .pop()
        .trim(' '); // EXAMPLE URL: /api/UserLogins?$filter=UserName eq 'isaac.hardy@newspring.cc'

      if (!identity) {
        const response = new Response('');
        response.status = 400;
        return Promise.reject(response);
      }
      if (identity === `'isaac.hardy@newspring.cc'`)
        return resolveWith([rockMocks.userLogins()]);

      return resolveWith([]);
    }
    console.log(`No route matching ${url}`);
    return Promise.reject(`No route matching ${url}`);
  });
};

const apolloServerEnvMocked = {
  ...apolloServerEnv,
  Response: fetch.Response,
  fetch,
};

module.exports = apolloServerEnvMocked;
