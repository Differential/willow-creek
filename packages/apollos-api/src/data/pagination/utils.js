// eslint-disable-next-line import/prefer-default-export
export const withEdgePagination = async ({ edges }) => {
  const result = await edges;
  const startCursor = result.length ? result[0].cursor : null;
  const endCursor = result.length ? result[result.length - 1].cursor : null;
  return { startCursor, endCursor };
};
