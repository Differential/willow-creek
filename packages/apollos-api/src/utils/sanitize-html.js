import sanitizeHtml from 'sanitize-html';

const allowedTags = [
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'blockquote',
  'p',
  'a',
  'ul',
  'ol',
  'li',
  'b',
  'i',
  'strong',
  'em',
  'br',
  'caption',
  'img',
];

const allowedAttributes = {
  a: ['href'],
  img: ['src'],
};

// A very picky HTML sanitizer
export default function(dirty) {
  return sanitizeHtml(dirty, {
    allowedTags,
    allowedAttributes,
  });
}
