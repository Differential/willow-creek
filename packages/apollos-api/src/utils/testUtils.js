import { KeyValueCache } from 'apollo-server-caching';
import {
  context as getContext,
  dataSources as getDataSources,
} from '../data/index';

export function getTestContext(req) {
  const context = getContext(req);
  const dataSources = getDataSources();
  // Apollo Server does this internally.
  Object.values(dataSources).forEach((dataSource) => {
    if (dataSource.initialize) {
      dataSource.initialize({ context, cache: KeyValueCache });
    }
  });
  context.dataSources = dataSources;
  return context;
}

export const buildGetMock = (response, dataSource) => {
  const get = jest.fn();
  if (Array.isArray(response) && Array.isArray(response[0])) {
    response.forEach((responseVal) => {
      get.mockReturnValueOnce(
        new Promise((resolve) => resolve(dataSource.normalize(responseVal)))
      );
    });
  }
  get.mockReturnValue(
    new Promise((resolve) => resolve(dataSource.normalize(response)))
  );
  return get;
};
