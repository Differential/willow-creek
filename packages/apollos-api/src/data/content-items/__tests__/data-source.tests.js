import { fetch } from 'apollo-server-env';
import ApollosConfig from '@apollosproject/config';
import { buildGetMock } from '../../../utils/testUtils';

import ContentItemsDataSource from '../data-source';

ApollosConfig.loadJs({
  ROCK: {
    API_URL: 'https://apollosrock.newspring.cc/api',
    API_TOKEN: 'some-rock-token',
    IMAGE_URL: 'https://apollosrock.newspring.cc/GetImage.ashx',
  },
});

describe('ContentItemsModel', () => {
  beforeEach(() => {
    fetch.resetMocks();
  });
  it('constructs', () => {
    expect(new ContentItemsDataSource()).toBeTruthy();
  });
  it('filters by content channel id', () => {
    const dataSource = new ContentItemsDataSource();
    dataSource.get = buildGetMock([{ Id: 1 }, { Id: 2 }], dataSource);
    const result = dataSource.byContentChannelId(1).get();
    expect(result).resolves.toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
  });

  it('gets by id', () => {
    const dataSource = new ContentItemsDataSource();
    dataSource.get = buildGetMock({ Id: 1 }, dataSource);
    const result = dataSource.getFromId(1);
    expect(result).resolves.toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
  });

  it('gets a cursor finding child content items of a provided parent', async () => {
    const dataSource = new ContentItemsDataSource();
    dataSource.get = buildGetMock(
      [
        [
          { ChildContentChannelItemId: 101 },
          { ChildContentChannelItemId: 201 },
        ],
        [{ Id: 1 }, { Id: 2 }],
      ],
      dataSource
    );
    const cursor = await dataSource.getCursorByParentContentItemId(1);
    expect(cursor.get()).resolves.toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
  });

  it('returns null when there are no child content items', async () => {
    const dataSource = new ContentItemsDataSource();
    dataSource.get = buildGetMock([], dataSource);
    const cursor = await dataSource.getCursorByParentContentItemId(1);
    expect(cursor).toBe(null);
    expect(dataSource.get.mock.calls).toMatchSnapshot();
  });

  it('gets a cursor finding parent content items of a provided child', async () => {
    const dataSource = new ContentItemsDataSource();
    dataSource.get = buildGetMock(
      [
        [{ ContentChannelItemId: 101 }, { ContentChannelItemId: 201 }],
        [{ Id: 1 }, { Id: 2 }],
      ],
      dataSource
    );
    const cursor = await dataSource.getCursorByChildContentItemId(1);
    expect(cursor.get()).resolves.toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
  });

  it('gets a cursor finding sibling content items of a provided item', async () => {
    const dataSource = new ContentItemsDataSource();
    dataSource.get = buildGetMock(
      [
        [{ ContentChannelItemId: 101 }],
        [
          { ContentChannelId: 201, ChildContentChannelItemId: 1 },
          { ContentChannelId: 202, ChildContentChannelItemId: 2 },
        ],
        [{ Id: 1 }, { Id: 2 }],
      ],
      dataSource
    );
    const cursor = await dataSource.getCursorBySiblingContentItemId(1);
    expect(cursor.get()).resolves.toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
  });

  it('returns null when there are no parent content items', async () => {
    const dataSource = new ContentItemsDataSource();
    dataSource.get = buildGetMock([], dataSource);
    const cursor = await dataSource.getCursorByChildContentItemId(1);
    expect(cursor).toBe(null);
    expect(dataSource.get.mock.calls).toMatchSnapshot();
  });
});
