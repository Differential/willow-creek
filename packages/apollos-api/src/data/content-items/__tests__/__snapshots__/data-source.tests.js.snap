// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentItemsModel filters by content channel id 1`] = `
Array [
  Array [
    "ContentChannelItems?loadAttributes=expanded&%24filter=ContentChannelId%20eq%201&%24orderby=StartDateTime%20desc",
    Object {},
    Object {},
  ],
]
`;

exports[`ContentItemsModel filters by content channel id 2`] = `
Array [
  Object {
    "id": 1,
  },
  Object {
    "id": 2,
  },
]
`;

exports[`ContentItemsModel gets a cursor finding child content items of a provided parent 1`] = `
Array [
  Array [
    "ContentChannelItemAssociations?loadAttributes=expanded&%24filter=ContentChannelItemId%20eq%201",
    Object {},
    Object {},
  ],
  Array [
    "ContentChannelItems?loadAttributes=expanded&%24filter=%28Id%20eq%20101%29%20or%20%28Id%20eq%20201%29&%24orderby=Order%20asc",
    Object {},
    Object {},
  ],
]
`;

exports[`ContentItemsModel gets a cursor finding child content items of a provided parent 2`] = `
Array [
  Object {
    "id": 1,
  },
  Object {
    "id": 2,
  },
]
`;

exports[`ContentItemsModel gets a cursor finding parent content items of a provided child 1`] = `
Array [
  Array [
    "ContentChannelItemAssociations?loadAttributes=expanded&%24filter=ChildContentChannelItemId%20eq%201",
    Object {},
    Object {},
  ],
  Array [
    "ContentChannelItems?loadAttributes=expanded&%24filter=%28Id%20eq%20101%29%20or%20%28Id%20eq%20201%29&%24orderby=Order%20asc",
    Object {},
    Object {},
  ],
]
`;

exports[`ContentItemsModel gets a cursor finding parent content items of a provided child 2`] = `
Array [
  Object {
    "id": 1,
  },
  Object {
    "id": 2,
  },
]
`;

exports[`ContentItemsModel gets a cursor finding sibling content items of a provided item 1`] = `
Array [
  Array [
    "ContentChannelItemAssociations?loadAttributes=expanded&%24filter=ChildContentChannelItemId%20eq%201",
    Object {},
    Object {},
  ],
  Array [
    "ContentChannelItemAssociations?loadAttributes=expanded&%24filter=%28ContentChannelItemId%20eq%20101%29%20and%20%28ChildContentChannelItemId%20ne%201%29",
    Object {},
    Object {},
  ],
  Array [
    "ContentChannelItems?loadAttributes=expanded&%24filter=%28Id%20eq%201%29%20or%20%28Id%20eq%202%29&%24orderby=Order%20asc",
    Object {},
    Object {},
  ],
]
`;

exports[`ContentItemsModel gets a cursor finding sibling content items of a provided item 2`] = `
Array [
  Object {
    "id": 1,
  },
  Object {
    "id": 2,
  },
]
`;

exports[`ContentItemsModel gets by id 1`] = `
Array [
  Array [
    "ContentChannelItems/1?loadAttributes=expanded",
    Object {},
    Object {},
  ],
]
`;

exports[`ContentItemsModel gets by id 2`] = `
Object {
  "id": 1,
}
`;

exports[`ContentItemsModel returns null when there are no child content items 1`] = `
Array [
  Array [
    "ContentChannelItemAssociations?loadAttributes=expanded&%24filter=ContentChannelItemId%20eq%201",
    Object {},
    Object {},
  ],
]
`;

exports[`ContentItemsModel returns null when there are no parent content items 1`] = `
Array [
  Array [
    "ContentChannelItemAssociations?loadAttributes=expanded&%24filter=ChildContentChannelItemId%20eq%201",
    Object {},
    Object {},
  ],
]
`;
